File: e24_nominals.py
class E24Nominals:
    E24 = (1.0, 1.1, 1.2, 1.3, 1.5, 1.6, 1.8, 2.0, 2.2, 2.4, 2.7, 3.0, 3.3,
           3.6, 3.9, 4.3, 4.7, 5.1, 5.6, 6.2, 6.8, 7.5, 8.2, 9.1)

    @staticmethod
    def calculate_standard_resistor(resistance, go_up: bool):
        if resistance >= 1:
            res = resistance
            while res >= 10:
                res /= 10
            list_of_diffs = [abs(res - x) for x in E24Nominals.E24]
            result_index = list_of_diffs.index(min(list_of_diffs))
            interact = resistance / E24Nominals.E24[result_index]
            power = 0
            while True:
                if interact < 9:
                    break
                power += 1
                interact /= 10
            e24_result = E24Nominals.E24[result_index] * 10 ** power
            if go_up:
                if e24_result < resistance:
                    if result_index != len(E24Nominals.E24) - 1:
                        e24_result = E24Nominals.E24[result_index + 1] * \
                            10 ** power
                    else:
                        e24_result = E24Nominals.E24[0] * 10 ** (power + 1)
        else:
            power = 0
            while resistance < 1:
                resistance *= 10
                power += 1
            list_of_difs = [abs(resistance - x) for x in E24Nominals.E24]
            result_index = list_of_difs.index(min(list_of_difs))

            e24_result = E24Nominals.E24[result_index] * 10 ** -power
        return e24_result

File: AndroidStudio/app/src/main/java/com/lemanrus/radioman/domain/E24Nominals.kt
package com.lemanrus.radioman.domain

object E24Nominals {
    private val E24 = doubleArrayOf(
        1.0, 1.1, 1.2, 1.3, 1.5, 1.6, 1.8, 2.0, 2.2, 2.4, 2.7, 3.0, 3.3,
        3.6, 3.9, 4.3, 4.7, 5.1, 5.6, 6.2, 6.8, 7.5, 8.2, 9.1
    )

    fun calculateStandardResistor(resistance: Double, goUp: Boolean): Double {
        if (resistance >= 1) {
            var res = resistance
            while (res >= 10) res /= 10.0
            val listOfDiffs = E24.map { kotlin.math.abs(res - it) }
            var resultIndex = listOfDiffs.indexOf(listOfDiffs.minOrNull()!!)
            var interact = resistance / E24[resultIndex]
            var power = 0
            while (interact >= 9) {
                power++
                interact /= 10.0
            }
            var e24Result = E24[resultIndex] * Math.pow(10.0, power.toDouble())
            if (goUp && e24Result < resistance) {
                e24Result = if (resultIndex != E24.lastIndex) {
                    E24[resultIndex + 1] * Math.pow(10.0, power.toDouble())
                } else {
                    E24[0] * Math.pow(10.0, (power + 1).toDouble())
                }
            }
            return e24Result
        } else {
            var r = resistance
            var power = 0
            while (r < 1) {
                r *= 10.0
                power++
            }
            val listOfDiffs = E24.map { kotlin.math.abs(r - it) }
            val resultIndex = listOfDiffs.indexOf(listOfDiffs.minOrNull()!!)
            return E24[resultIndex] * Math.pow(10.0, -power.toDouble())
        }
    }
}
